platform/saivpp/vpplib/SwitchStateBaseFdb.cpp

```diff
@@ -1082,3 +1159,237 @@ sai_status_t SwitchStateBase::vpp_delete_bvi_interface(
+        auto attr = tunnel_attrs[meta->attridname];
+        if (attr == NULL) {
+            SWSS_LOG_ERROR("SAI_OBJECT_TYPE_TUNNEL 0x%lx has no SAI_TUNNEL_ATTR_PEER_MODE",
+                    sai_serialize_object_id(tunnel_id).c_str());
+            return SAI_STATUS_FAILURE;
+        }
+
+        if (attr->getAttr()->value.s32 != SAI_TUNNEL_PEER_MODE_P2P) {
+            SWSS_LOG_ERROR("Only SAI_TUNNEL_PEER_MODE_P2P supported", sai_serialize_object_id(tunnel_id).c_str());
+            return SAI_STATUS_FAILURE;
+        }
+    }
+
+    // SIP
+    meta = sai_metadata_get_attr_metadata(SAI_OBJECT_TYPE_TUNNEL, SAI_TUNNEL_ATTR_ENCAP_SRC_IP);
+    sip_attr = tunnel_attrs[meta->attridname]->getAttr();
+    if (sip_attr == NULL) {
+        SWSS_LOG_ERROR("SAI_OBJECT_TYPE_TUNNEL 0x%lx has no SAI_TUNNEL_ATTR_ENCAP_SRC_IP",
+                sai_serialize_object_id(tunnel_id).c_str());
+        return SAI_STATUS_FAILURE;
+    }
+
+    // DIP
+    meta = sai_metadata_get_attr_metadata(SAI_OBJECT_TYPE_TUNNEL, SAI_TUNNEL_ATTR_ENCAP_DST_IP);
+    dip_attr = tunnel_attrs[meta->attridname]->getAttr();
+    if (dip_attr == NULL) {
+        SWSS_LOG_ERROR("SAI_OBJECT_TYPE_TUNNEL 0x%lx has no SAI_TUNNEL_ATTR_ENCAP_DST_IP",
+                sai_serialize_object_id(tunnel_id).c_str());
+        return SAI_STATUS_FAILURE;
+    }
+
+    // DECAP_MAPPERS
+    meta = sai_metadata_get_attr_metadata(SAI_OBJECT_TYPE_TUNNEL, SAI_TUNNEL_ATTR_DECAP_MAPPERS);
+    decap_attr = tunnel_attrs[meta->attridname]->getAttr();
+
+    // VLAN
+    auto attr_vlanid_map = m_objectHash.at(SAI_OBJECT_TYPE_VLAN).at(sai_serialize_object_id(vlan_oid));
+    auto md_vlan_id = sai_metadata_get_attr_metadata(SAI_OBJECT_TYPE_VLAN, SAI_VLAN_ATTR_VLAN_ID);
+    vlan = attr_vlanid_map.at(md_vlan_id->attridname)->getAttr()->value.u16;
+
+    // VNI lookup
+    if (vpp_find_vni(vlan, decap_attr, &vni_attr) != SAI_STATUS_SUCCESS ){
+        SWSS_LOG_DEBUG("No VNI found for VLAN %hd", vlan);
+        return SAI_STATUS_FAILURE;
+    }
+
+    sai_status_t status = vpp_create_vxlan_tunnel(sip_attr, dip_attr, vni_attr, if_ind);
+    if (status == SAI_STATUS_SUCCESS)
+        m_tunnel_oid_to_iface_map[tunnel_id] = *if_ind;
+
+    return status;
+}
+
+
+/**
+  Lookup for VNI in TUNNEL_MAP_ENTRY's
+
+  Example of  TUNNEL_MAP_ENTRY:
+
+    vpp_create_tunnel_map_entry(
+      SAI_OBJECT_TYPE_TUNNEL_MAP_ENTRY
+        #0 SAI_TUNNEL_MAP_ENTRY_ATTR_TUNNEL_MAP_TYPE=VNI_TO_VLAN_ID
+        #1 SAI_TUNNEL_MAP_ENTRY_ATTR_TUNNEL_MAP=0x2900000000
+        #2 SAI_TUNNEL_MAP_ENTRY_ATTR_VLAN_ID_VALUE=100
+        #3 SAI_TUNNEL_MAP_ENTRY_ATTR_VNI_ID_KEY=1000
+    ) = 0x3b00000000
+*/
+
+sai_status_t SwitchStateBase::vpp_find_vni(
+        _In_  sai_uint16_t vlan,
+        _In_  const sai_attribute_t *decap_attr,
+        _Out_ const sai_attribute_t **vni_attr)
+{
+    for (const auto &obj_it : m_objectHash.at(SAI_OBJECT_TYPE_TUNNEL_MAP_ENTRY)) {
```

Comment:
This is very inefficient to go through all tunnel_map_entries. In #97, I have the same need to find tunnel_map_entries from tunnel_map. To efficiently process that, I introduced SaiObjectDB, which dynamically build the relationship on demand. Please take a look.

Reply:
m_objectHash is the DB for SAI objects. If you are going to create another DB for SAI objects then I would say this is not a great idea. 
For me the rule of thumb is to avoid at all costs keeping same data in more then one place. Otherwise you are responsible for keeping the data in sync between different places which significantly complicates the implementation.
The other reason not to create another DB is the current implementation. If you have a look at the code you'll not find an example of using another DB for particular feature.

Reply:
It's trade off. The more code you add the more complex. But this is necessary.
1. SAI objects are basically a graph. Using a flat hash table is not appropriate. With proper graph data structure, we can improve navigation efficiency for the situation you have.
2. You can see how many customized map to maintain the relationship: m_acl_tbl_rules_map, m_acl_tbl_grp_mbr_map, m_nxthop_grp_mbr_map .... The goal of SaiObjectDB is to eliminate those maps with common and standard implementation. So we don't need to repeat adding such code that will increase the risk of adding more bugs.
3. Using proper OO methodology can improve code readability and modularity. 

Adding some infra has the risk of short term instability but the benefit pays off in the long run.